pkg/ottl/contexts/internal/ctxutil/slice.go

```diff
@@ -5,61 +5,258 @@ package ctxutil // import "github.com/open-telemetry/opentelemetry-collector-con
+	"errors"
 	"fmt"
 
 	"go.opentelemetry.io/collector/pdata/pcommon"
+	"golang.org/x/exp/constraints"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
 )
 
-func GetSliceValue[K any](ctx context.Context, tCtx K, s pcommon.Slice, keys []ottl.Key[K]) (any, error) {
+const (
+	typeNotIndexableError = "type %T does not support indexing"
+)
+
+var (
+	errMissingSetKey = errors.New("cannot set slice value without key")
+	errMissingGetKey = errors.New("cannot get slice value without key")
+)
+
+func getSliceIndexFromKeys[K any](ctx context.Context, tCtx K, sliceLen int, keys []ottl.Key[K]) (int, error) {
 	if len(keys) == 0 {
-		return nil, fmt.Errorf("cannot get slice value without key")
+		return 0, fmt.Errorf("cannot get/set slice index without a key")
 	}
 
 	i, err := keys[0].Int(ctx, tCtx)
 	if err != nil {
-		return nil, err
+		return 0, err
 	}
 	if i == nil {
 		resInt, err := FetchValueFromExpression[K, int64](ctx, tCtx, keys[0])
 		if err != nil {
-			return nil, fmt.Errorf("unable to resolve an integer index in slice: %w", err)
+			return 0, fmt.Errorf("unable to resolve an integer index in slice: %w", err)
 		}
 		i = resInt
 	}
 
 	idx := int(*i)
 
-	if idx < 0 || idx >= s.Len() {
-		return nil, fmt.Errorf("index %d out of bounds", idx)
+	if idx < 0 || idx >= sliceLen {
+		return 0, fmt.Errorf("index %d out of bounds", idx)
+	}
+
+	return idx, nil
+}
+
+func GetSliceValue[K any](ctx context.Context, tCtx K, s pcommon.Slice, keys []ottl.Key[K]) (any, error) {
+	if len(keys) == 0 {
+		return 0, errMissingGetKey
+	}
+
+	idx, err := getSliceIndexFromKeys(ctx, tCtx, s.Len(), keys)
+	if err != nil {
+		return nil, err
 	}
 
 	return getIndexableValue[K](ctx, tCtx, s.At(idx), keys[1:])
 }
 
 func SetSliceValue[K any](ctx context.Context, tCtx K, s pcommon.Slice, keys []ottl.Key[K], val any) error {
 	if len(keys) == 0 {
-		return fmt.Errorf("cannot set slice value without key")
+		return errMissingSetKey
 	}
 
-	i, err := keys[0].Int(ctx, tCtx)
+	idx, err := getSliceIndexFromKeys(ctx, tCtx, s.Len(), keys)
 	if err != nil {
 		return err
 	}
-	if i == nil {
-		resInt, err := FetchValueFromExpression[K, int64](ctx, tCtx, keys[0])
-		if err != nil {
-			return fmt.Errorf("unable to resolve an integer index in slice: %w", err)
+
+	return SetIndexableValue[K](ctx, tCtx, s.At(idx), val, keys[1:])
+}
+
+// CommonTypedSlice is an interface for typed pdata slices, such as pcommon.StringSlice,
+// pcommon.Int64Slice, pcommon.Int32Slice, etc.
+type CommonTypedSlice[T any] interface {
+	At(int) T
+	Len() int
+	FromRaw(val []T)
+	SetAt(int, T)
+	AsRaw() []T
+}
+
+// GetCommonTypedSliceValue is like GetSliceValue, but for retrieving a value from a pdata
+// typed slice. [V] is the type of the slice elements. If no keys are provided, it returns
+// an error. This function does not support slice elements indexing.
+func GetCommonTypedSliceValue[K, V any](ctx context.Context, tCtx K, s CommonTypedSlice[V], keys []ottl.Key[K]) (V, error) {
```

Comment:
I'd like to understand why `keys` is an array when at runtime there is a check for only one element. IMO it would be better if we can swap the runtime check by a compiler check (`key ottl.Key[K]`), but I likely miss some context here.